src/xds/provider/agones.rs

```diff
@@ -0,0 +1,438 @@
+        let resources = match kind {
+            ResourceType::Endpoint => {
+                let cluster = self.get_cluster().await?;
+                let mut buf = Vec::new();
+                buf.reserve(cluster.encoded_len());
+
+                cluster.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: crate::xds::ENDPOINT_TYPE.into(),
+                    value: buf,
+                }]
+            }
+            ResourceType::Listener => {
+                let listener = self.get_listener().await?;
+                let mut buf = Vec::new();
+                buf.reserve(listener.encoded_len());
+
+                listener.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: kind.type_url().into(),
+                    value: buf,
+                }]
+            }
+            kind => {
+                return Err(tonic::Status::internal(format!(
+                    "Quilkin currently does not support {} requests",
+                    kind.type_url()
+                )))
+            }
+        };
+
+        Ok(DiscoveryResponse {
+            version_info: version.to_string(),
+            resources,
+            type_url: kind.type_url().into(),
+            ..<_>::default()
+        })
+    }
+}
+
+#[derive(kube::CustomResource, Clone, Debug, Deserialize, Serialize, JsonSchema)]
+#[kube(
+    group = "agones.dev",
+    version = "v1",
+    kind = "GameServer",
+    status = "GameServerStatus",
+    namespaced
+)]
+struct GameServerSpec {
+    /// Container specifies which Pod container is the game server. Only
+    /// required if there is more than one container defined.
+    container: String,
+    /// Ports are the array of ports that can be exposed via the game server
+    ports: Vec<GameServerPort>,
+    /// Configures health checking
+    health: Health,
+    /// Scheduling strategy. Defaults to "Packed"
+    scheduling: SchedulingStrategy,
+    /// Specifies parameters for the Agones SDK Server sidecar container.
+    #[serde(rename = "sdkServer")]
+    sdk_server: SdkServer,
+    /// Describes the Pod that will be created for the [`GameServer`].
+    template: k8s_openapi::api::core::v1::PodTemplateSpec,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+struct Health {
+    /// Whether health checking is disabled or not
+    #[serde(default)]
+    disabled: bool,
+    /// The number of seconds each health ping has to occur in
+    #[serde(rename = "periodSeconds", default = "default_period_seconds")]
+    period_seconds: i32,
+    /// How many failures in a row constitutes unhealthy
+    #[serde(rename = "failureThreshold", default = "default_failure_threshold")]
+    failure_threshold: i32,
+    /// Initial delay before checking health
+    #[serde(
+        rename = "initialDelaySeconds",
+        default = "default_initial_delay_seconds"
+    )]
+    initial_delay_seconds: i32,
+}
+
+fn default_period_seconds() -> i32 {
```

Comment:
I feel like you shouldn't need all of this default value logic (or keep them maintained in case any of them change). The default values would be set at the Agones controller/k8s layer, so you could potentially remove all of these and cut down on some long term maintenance costs.

Reply:
I thought so too, but when I tried connecting to our clusters we hit some of these fields being absent (they are marked with `omitempty` in Go), and thus I need to provide some default value for them.

Reply:
Aaaah. That makes sense. It's not like the defaults are going to change on the Agones side, so it is going to be implemented once (🤞🏻)